chromium_src/url/url_canon_stdurl.cc

```diff
@@ -0,0 +1,127 @@
+
+std::string Base32Encode(base::StringPiece input) {
+  if (input.empty())
+    return std::string();
+
+  // When no padding is used, the output is exactly 1 character per 5 bits of
+  // input and one more for the last [1,4] bits.
+  // That is: ceil(input.size() * 8.0 / 5.0) ==
+  //          (input.size() * 8 + 4) / 5.
+  const size_t unpadded_length =
+      ((base::MakeCheckedNum(input.size()) * 8 + 4) / 5).ValueOrDie();
+
+  std::string output;
+  const size_t encoded_length = unpadded_length;
+  output.reserve(encoded_length);
+
+  // A bit stream which will be read from the left and appended to from the
+  // right as it's emptied.
+  uint16_t bit_stream = (static_cast<uint8_t>(input[0]) << 8);
+  size_t next_byte_index = 1;
+  int free_bits = 8;
+  while (free_bits < 16) {
+    // Extract the 5 leftmost bits in the stream
+    output.push_back(kEncoding[bit_stream >> 11]);
+    bit_stream <<= 5;
+    free_bits += 5;
+
+    // If there is enough room in the bit stream, inject another byte (if there
+    // are any left...).
+    if (free_bits >= 8 && next_byte_index < input.size()) {
+      free_bits -= 8;
+      bit_stream += static_cast<uint8_t>(input[next_byte_index++]) << free_bits;
+    }
+  }
+
+  DCHECK_EQ(encoded_length, output.size());
+  return output;
+}
+
+std::string CIDV0ToV1(const std::string& v0) {
+  if (!base::StartsWith(v0, kCIDv0Prefix) || v0.size() != 46)
+    return std::string();
+  std::vector<unsigned char> buf(34);
+  if (!DecodeBase58(v0.c_str(), buf, buf.size())) {
+    return std::string();
+  }
+  std::string hash;
+  hash.insert(hash.end(), buf.begin(), buf.end());
+  std::string codec;
+  codec.insert(0, 1, static_cast<char>(0x70));
+  std::string version;
+  version.insert(0, 1, static_cast<char>(1));
+  // <multicodec-cidv1><multicodec-content-type><multihash-content-address>
+  base::StringPiece input(version + codec + hash);
+  std::string encoded = Base32Encode(input);
+  if (encoded.empty())
+    return std::string();
+  std::string trimmed;
+  base::TrimString(encoded, "=", &trimmed);
+  // multibase format <base-encoding-character><base-encoded-data>
+  // https://github.com/multiformats/multibase/blob/master/multibase.csv
+  std::string lowercase = "b" + base::ToLowerASCII(trimmed);
+  return lowercase;
+}
+}  // namespace
+
+namespace ipfs {
+template <typename CHAR>
+bool IpfsCIDv0(const url::URLComponentSource<CHAR>& source,
+               const url::Parsed& parsed,
+               url::CanonOutput* output,
+               url::Parsed* new_parsed);
+}  // namespace ipfs
+
+#define CanonicalizeHost \
+  ipfs::IpfsCIDv0<CHAR>(source, parsed, output, new_parsed) || CanonicalizeHost
+
+#include "../../url/url_canon_stdurl.cc"
+#undef CanonicalizeHost
+
+namespace ipfs {
+template <typename CHAR>
+// Do not canonicalize CIDv0(Qm...) and convert it to CIDv1
+// https://docs.ipfs.io/concepts/content-addressing/#version-0-v0
+bool IpfsCIDv0(const url::URLComponentSource<CHAR>& source,

```

Comment:
are there tests for this?

Reply:
it covered directly by `IpfsUtilsUnitTest.ParseIPFSUri` and `IpfsUtilsUnitTest.IpfsUriReplacement` and indirectly by many others ipfs related